### PR TITLE
Fix and modernize lightning-invoice API

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Fixes https://github.com/lightningdevkit/ldk-garbagecollected/issues/41 and requires payment_secrets

### CODE REVIEW

**Review by jkczyz — APPROVED** [review]

File: lightning-invoice/src/lib.rs [hunk 1]
```diff
@@ -480,8 +485,9 @@ impl<D: tb::Bool, H: tb::Bool, T: tb::Bool, C: tb::Bool, S: tb::Bool> InvoiceBui
 		}
 	}
 
-	/// Sets the amount in pico BTC. The optimal SI prefix is choosen automatically.
-	pub fn amount_pico_btc(mut self, amount: u64) -> Self {
+	/// Sets the amount in millisatoshis. The optimal SI prefix is choosen automatically.
```
Comment (KaiWitt) [inline lightning-invoice/src/lib.rs:488]: *chosen

File: lightning-invoice/src/lib.rs [hunk 2]
```diff
@@ -1016,35 +1023,56 @@ impl Invoice {
...
+		// "A writer MUST include exactly one `s` field."
+		let payment_secret_count = self.tagged_fields().filter(|&tf| match *tf {
+			TaggedField::PaymentSecret(_) => true,
+			_ => false,
+		}).count();
+		if payment_secret_count < 1 {
+			return Err(SemanticError::NoPaymentSecret);
+		} else if payment_secret_count > 1 {
+			return Err(SemanticError::MultiplePaymentSecrets);
+		}
```
Comment (KaiWitt) [inline lightning-invoice/src/lib.rs:1035]: Maybe put this in a separate method, it's the same as in line 1053+

**Review by valentinewallace — APPROVED** [review]
LGTM

**Comment by TheBlueMatt** [comment]
Squashed without changes, will take after CI:

```
$ git diff-tree -U1 a5c58482 68793485
$
```

**Pull request merged**